File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root: return True
        def dfs(node):
            if node is None: return 0
            left, right = dfs(node.left), dfs(node.right)
            
            if left == -1 or right == -1 or abs(left - right) > 1:
                return -1
            
            return 1 + max(left, right)
        return True if dfs(root) != -1 else False

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#define pb push_back
class Solution {
public:
    int res = 0;
    void dfs(TreeNode* node, int maxSoFar) {
        if (!node) return;

        if (node->val >= maxSoFar) {
            res++;
            maxSoFar = node->val;
        }
        dfs(node->left, maxSoFar);
        dfs(node->right, maxSoFar);
    }

    int goodNodes(TreeNode* root) {
        if (!root) return 0;
        dfs(root, root->val);
        return res;
    }
};